modules/helpers.py

```diff
@@ -54,6 +57,56 @@ def next_inputs(self, time, outputs, state, sample_ids, name=None):
         return (finished, next_inputs, state)
 
 
+class OneHotValidationHelper(Helper):
+
+    def __init__(self, targets, batch_size, output_dim, r, n_feed_frame=1, teacher_forcing=False):
+        assert n_feed_frame <= r
+        self._batch_size = batch_size
+        self._output_dim = output_dim
+        self.r = r
+        self._end_token = tf.tile([0.0], [output_dim * r])
+        self.n_feed_frame = n_feed_frame
+        self.num_steps = tf.shape(targets)[1] // r
+        self.teacher_forcing = teacher_forcing
+        self._targets = tf.reshape(targets,
+                                   shape=tf.stack([self.batch_size, self.num_steps, tf.to_int32(output_dim * r)]))
+
+    @property
+    def batch_size(self):
+        return self._batch_size
+
+    @property
+    def sample_ids_shape(self):
+        return tf.TensorShape([self.r])
+
+    @property
+    def sample_ids_dtype(self):
+        return tf.int32
+
+    def initialize(self, name=None):
+        return (
+            tf.tile([False], [self._batch_size]),
+            _go_frames(self._batch_size, self._output_dim * self.n_feed_frame, self._targets.dtype))
+
+    def sample(self, time, outputs, state, name=None):
+        outputs, done = outputs
+        undo_reduction = tf.reshape(outputs, shape=[self._batch_size, self.r, self._output_dim])
+        samples = tf.argmax(undo_reduction, axis=-1, output_type=tf.int32)
+        return samples
+
+    def next_inputs(self, time, outputs, state, sample_ids, name=None):
+        output, done = outputs
+        undo_reduction = tf.reshape(output, shape=[self._batch_size, self.r, self._output_dim])
+        predicted_probs = tf.reshape(tf.nn.softmax(undo_reduction, axis=-1),
+                               shape=[self._batch_size, self.r * self._output_dim])
+        finished = (time + 1 >= self.num_steps)
+        next_inputs = self._targets[:, time,
```

Comment:
If teacher_forcing=True, a target is fed back in an autoregressive loop. If not, predicted probabilities are fed back (you have to modify this part if you want to feedback samples instead of probabilities).